### PR TITLE
Document some undocumented items

### PR DESCRIPTION
**Pull request by nw0**
This PR mainly focuses on the currently undocumented items (#306).

I've also added an `#[allow(missing_docs)]` to the `ffi` module.

The outstanding items seem to mainly be the protocols and exceptions macros, but I'm not very sure as my rustdoc reports 459 remaining items.

I think the exceptions macros need some love: will update the issue if and when I think of something reasonable.

### CODE REVIEW

**Review by messense — APPROVED** [review]
LGTM, thanks!

File: src/buffer.rs [hunk 1]
```diff
@@ -599,11 +606,13 @@ impl<T> Drop for PyBuffer<T> {
 pub struct ReadOnlyCell<T: Element>(cell::UnsafeCell<T>);
 
 impl<T: Element> ReadOnlyCell<T> {
+    /// Obtain a copy of the current content.
```
Comment (davidhewitt) [inline src/buffer.rs:609]:
```suggestion
    /// Returns a copy of the current value.
```

File: src/buffer.rs [hunk 2]
```diff
@@ -599,11 +606,13 @@ impl<T> Drop for PyBuffer<T> {
...
     #[inline]
     pub fn get(&self) -> T {
         unsafe { *self.0.get() }
     }
 
+    /// Get a pointer to the current content.
```
Comment (davidhewitt) [inline src/buffer.rs:615]:
```suggestion
    /// Returns a pointer to the current value.
```

File: src/exceptions.rs [hunk 3]
```diff
@@ -18,6 +18,7 @@ macro_rules! impl_exception_boilerplate {
         }
 
         impl $name {
+            /// Create a new [PyErr](crate::PyErr) of this type.
```
Comment (davidhewitt) [inline src/exceptions.rs:21]:
```suggestion
            /// Creates a new [PyErr](crate::PyErr) of this type.
```

File: src/exceptions.rs [hunk 4]
```diff
@@ -275,6 +276,7 @@ impl_native_exception!(PyIOError, PyExc_IOError);
 impl_native_exception!(PyWindowsError, PyExc_WindowsError);
 
 impl PyUnicodeDecodeError {
+    /// Create a Python `UnicodeDecodeError`.
```
Comment (davidhewitt) [inline src/exceptions.rs:279]:
```suggestion
    /// Creates a Python `UnicodeDecodeError`.
```

File: src/exceptions.rs [hunk 5]
```diff
@@ -294,6 +296,7 @@ impl PyUnicodeDecodeError {
         }
     }
 
+    /// Create a Python `UnicodeDecodeError` from a Rust UTF-8 decoding error.
```
Comment (davidhewitt) [inline src/exceptions.rs:299]:
```suggestion
    /// Creates a Python `UnicodeDecodeError` from a Rust UTF-8 decoding error.
```

File: src/instance.rs [hunk 6]
```diff
@@ -18,6 +18,7 @@ use std::ptr::NonNull;
 /// to the GIL, which is why you can get a token from all references of those
 /// types.
 pub unsafe trait PyNativeType: Sized {
+    /// Implicitly acquire the GIL marker.
```
Comment (davidhewitt) [inline src/instance.rs:21]:
```suggestion
    /// Returns a GIL marker constrained to the lifetime of this type.
```

File: src/types/datetime.rs [hunk 7]
```diff
@@ -29,12 +29,24 @@ use std::os::raw::c_int;
 #[cfg(not(PyPy))]
 use std::ptr;
 
-/// Access traits
+// Access traits
 
 /// Trait for accessing the date components of a struct containing a date.
 pub trait PyDateAccess {
+    /// Return the year, as a positive int.
```
Comment (davidhewitt) [inline src/types/datetime.rs:36]:
```suggestion
    /// Returns the year, as a positive int.
```

File: src/types/datetime.rs [hunk 8]
```diff
@@ -29,12 +29,24 @@ use std::os::raw::c_int;
...
+    ///
+    /// Implementations should conform to the upstream documentation:
+    /// <https://docs.python.org/3/c-api/datetime.html#c.PyDateTime_GET_YEAR>
     fn get_year(&self) -> i32;
+    /// Return the month, as an int from 1 through 12.
```
Comment (davidhewitt) [inline src/types/datetime.rs:41]:
```suggestion
    /// Returns the month, as an int from 1 through 12.
```

File: src/types/datetime.rs [hunk 9]
```diff
@@ -29,12 +29,24 @@ use std::os::raw::c_int;
...
+    ///
+    /// Implementations should conform to the upstream documentation:
+    /// <https://docs.python.org/3/c-api/datetime.html#c.PyDateTime_GET_MONTH>
     fn get_month(&self) -> u8;
+    /// Return the day, as an int from 1 through 31.
```
Comment (davidhewitt) [inline src/types/datetime.rs:46]:
```suggestion
    /// Returns the day, as an int from 1 through 31.
```

File: src/types/datetime.rs [hunk 10]
```diff
@@ -44,17 +56,51 @@ pub trait PyDateAccess {
 /// microsecond) representation of the delta, they are *not* intended as
 /// aliases for calculating the total duration in each of these units.
 pub trait PyDeltaAccess {
+    /// Return the number of days, as an int from -999999999 to 999999999.
```
Comment (davidhewitt) [inline src/types/datetime.rs:59]:
```suggestion
    /// Returns the number of days, as an int from -999999999 to 999999999.
```

File: src/types/datetime.rs [hunk 11]
```diff
@@ -44,17 +56,51 @@ pub trait PyDateAccess {
...
+    ///
+    /// Implementations should conform to the upstream documentation:
+    /// <https://docs.python.org/3/c-api/datetime.html#c.PyDateTime_DELTA_GET_DAYS>
     fn get_days(&self) -> i32;
+    /// Return the number of seconds, as an int from 0 through 86399.
```
Comment (davidhewitt) [inline src/types/datetime.rs:64]:
```suggestion
    /// Returns the number of seconds, as an int from 0 through 86399.
```

File: src/types/datetime.rs [hunk 12]
```diff
@@ -44,17 +56,51 @@ pub trait PyDateAccess {
...
+    ///
+    /// Implementations should conform to the upstream documentation:
+    /// <https://docs.python.org/3/c-api/datetime.html#c.PyDateTime_DELTA_GET_DAYS>
     fn get_seconds(&self) -> i32;
+    /// Return the number of microseconds, as an int from 0 through 999999.
```
Comment (davidhewitt) [inline src/types/datetime.rs:69]:
```suggestion
    /// Returns the number of microseconds, as an int from 0 through 999999.
```

File: src/types/datetime.rs [hunk 13]
```diff
@@ -44,17 +56,51 @@ pub trait PyDateAccess {
...
+    ///
+    /// Implementations should conform to the upstream documentation:
+    /// <https://docs.python.org/3/c-api/datetime.html#c.PyDateTime_DELTA_GET_DAYS>
     fn get_microseconds(&self) -> i32;
 }
 
 /// Trait for accessing the time components of a struct containing a time.
 pub trait PyTimeAccess {
+    /// Return the hour, as an int from 0 through 23.
```
Comment (davidhewitt) [inline src/types/datetime.rs:78]:
```suggestion
    /// Returns the hour, as an int from 0 through 23.
```

File: src/types/datetime.rs [hunk 14]
```diff
@@ -44,17 +56,51 @@ pub trait PyDateAccess {
...
+    ///
+    /// Implementations should conform to the upstream documentation:
+    /// <https://docs.python.org/3/c-api/datetime.html#c.PyDateTime_DATE_GET_HOUR>
     fn get_hour(&self) -> u8;
+    /// Return the minute, as an int from 0 through 59.
```
Comment (davidhewitt) [inline src/types/datetime.rs:83]:
```suggestion
    /// Returns the minute, as an int from 0 through 59.
```

File: src/types/datetime.rs [hunk 15]
```diff
@@ -44,17 +56,51 @@ pub trait PyDateAccess {
...
+    ///
+    /// Implementations should conform to the upstream documentation:
+    /// <https://docs.python.org/3/c-api/datetime.html#c.PyDateTime_DATE_GET_MINUTE>
     fn get_minute(&self) -> u8;
+    /// Return the second, as an int from 0 through 59.
```
Comment (davidhewitt) [inline src/types/datetime.rs:88]:
```suggestion
    /// Returns the second, as an int from 0 through 59.
```

File: src/types/datetime.rs [hunk 16]
```diff
@@ -44,17 +56,51 @@ pub trait PyDateAccess {
...
+    ///
+    /// Implementations should conform to the upstream documentation:
+    /// <https://docs.python.org/3/c-api/datetime.html#c.PyDateTime_DATE_GET_SECOND>
     fn get_second(&self) -> u8;
+    /// Return the microsecond, as an int from 0 through 999999.
```
Comment (davidhewitt) [inline src/types/datetime.rs:93]:
```suggestion
    /// Returns the microsecond, as an int from 0 through 999999.
```

File: src/types/datetime.rs [hunk 17]
```diff
@@ -44,17 +56,51 @@ pub trait PyDateAccess {
...
+    ///
+    /// Implementations should conform to the upstream documentation:
+    /// <https://docs.python.org/3/c-api/datetime.html#c.PyDateTime_DATE_GET_MICROSECOND>
     fn get_microsecond(&self) -> u32;
+    /// Return whether this date is the later of two moments with the
```
Comment (davidhewitt) [inline src/types/datetime.rs:98]:
```suggestion
    /// Returns whether this date is the later of two moments with the
```

File: src/types/datetime.rs [hunk 18]
```diff
@@ -71,6 +117,7 @@ pyobject_native_type!(
 );
 
 impl PyDate {
+    /// Create a new `datetime.date`.
```
Comment (davidhewitt) [inline src/types/datetime.rs:120]:
```suggestion
    /// Creates a new `datetime.date`.
```

File: src/types/datetime.rs [hunk 19]
```diff
@@ -273,6 +320,7 @@ pyobject_native_type!(
 );
 
 impl PyTime {
+    /// Create a new `datetime.time` object.
```
Comment (davidhewitt) [inline src/types/datetime.rs:323]:
```suggestion
    /// Creates a new `datetime.time` object.
```

File: src/types/datetime.rs [hunk 20]
```diff
@@ -368,6 +416,7 @@ pyobject_native_type!(
 );
 
 impl PyDelta {
+    /// Create a new `timedelta`.
```
Comment (davidhewitt) [inline src/types/datetime.rs:419]:
```suggestion
    /// Creates a new `timedelta`.
```

File: src/types/sequence.rs [hunk 21]
```diff
@@ -28,6 +28,7 @@ impl PySequence {
         }
     }
 
+    /// Whether the sequence is empty.
```
Comment (davidhewitt) [inline src/types/sequence.rs:31]:
```suggestion
    /// Returns whether the sequence is empty.
```

File: src/types/string.rs [hunk 22]
```diff
@@ -27,6 +27,7 @@ impl PyString {
         unsafe { py.from_owned_ptr(ffi::PyUnicode_FromStringAndSize(ptr, len)) }
     }
 
+    /// Attempt to create a Python string from a Python bytes-like object.
```
Comment (davidhewitt) [inline src/types/string.rs:30]:
```suggestion
    /// Attempts to create a Python string from a Python [bytes-like object].
    ///
    /// [bytes-like object]: (https://docs.python.org/3/glossary.html#term-bytes-like-object).
```

**Review by davidhewitt — COMMENTED** [review]
Thank you very much for it, it's much appreciated!

I only have one nit, which is that I've changed the tense used e.g. "Return a pointer" to "Returns a pointer". 

This is the same tense used by Rust's stdlib and I was thinking it's nice to use the same style here. Admittedly this is undocumented, we could perhaps add a docs style note somewhere in `Contributing.md`....

File: src/buffer.rs [hunk 23]
```diff
@@ -55,16 +55,23 @@ impl<T> Debug for PyBuffer<T> {
...
+    /// A boolean type.
     Bool,
+    /// A float type and its width in bytes.
     Float { bytes: usize },
+    /// An unknown type. This may occur when parsing has failed.
     Unknown,
 }
 
 impl ElementType {
+    /// Determine the `ElementType` from a Python `struct` module format string.
```
Comment (birkenfeld) [inline src/buffer.rs:74]:
```suggestion
    /// Determines the `ElementType` from a Python `struct` module format string.
```

File: src/ffi/mod.rs [hunk 24]
```diff
@@ -4,6 +4,10 @@
 //! It is meant for advanced users only - regular PyO3 users shouldn't
 //! need to interact with this module at all.
 //!
+//! The contents of this module are not documented here, as it would entail
+//! keeping the documentation up to date. Consult the [Python/C API Reference
```
Comment (birkenfeld) [inline src/ffi/mod.rs:8]: I know what you mean, but it reads a bit funny - of course documentation always has to be kept up to date :)

```suggestion
//! basically copying the documentation from CPython. Consult the [Python/C API Reference
```

File: src/pyclass_slots.rs [hunk 25]
```diff
@@ -4,17 +4,24 @@ use crate::{ffi, Python};
 
 /// Represents `__dict__` field for `#[pyclass]`.
 pub trait PyClassDict {
+    /// Whether this `__dict__` field is capable of holding a dictionary.
     const IS_DUMMY: bool = true;
+    /// Initialise a [PyObject](crate::ffi::PyObject) `__dict__` reference.
```
Comment (birkenfeld) [inline src/pyclass_slots.rs:9]:
```suggestion
    /// Initializes a [PyObject](crate::ffi::PyObject) `__dict__` reference.
```

File: src/pyclass_slots.rs [hunk 26]
```diff
@@ -4,17 +4,24 @@ use crate::{ffi, Python};
...
     fn new() -> Self;
+    /// Empty the dictionary of its key-value pairs.
```
Comment (birkenfeld) [inline src/pyclass_slots.rs:11]:
```suggestion
    /// Empties the dictionary of its key-value pairs.
```

File: src/pyclass_slots.rs [hunk 27]
```diff
@@ -4,17 +4,24 @@ use crate::{ffi, Python};
...
     private_decl! {}
 }
 
 /// Represents `__weakref__` field for `#[pyclass]`.
 pub trait PyClassWeakRef {
+    /// Whether this `weakref` type is capable of holding weak references.
     const IS_DUMMY: bool = true;
+    /// Initialize a `weakref` instance.
     fn new() -> Self;
+    /// Clear the weak references to the given object.
```
Comment (birkenfeld) [inline src/pyclass_slots.rs:23]:
```suggestion
    /// Clears the weak references to the given object.
```

File: src/pyclass_slots.rs [hunk 28]
```diff
@@ -4,17 +4,24 @@ use crate::{ffi, Python};
...
     #[inline]
     fn clear_dict(&mut self, _py: Python) {}
     private_decl! {}
 }
 
 /// Represents `__weakref__` field for `#[pyclass]`.
 pub trait PyClassWeakRef {
+    /// Whether this `weakref` type is capable of holding weak references.
     const IS_DUMMY: bool = true;
+    /// Initialize a `weakref` instance.
```
Comment (birkenfeld) [inline src/pyclass_slots.rs:21]:
```suggestion
    /// Initializes a `weakref` instance.
```

**Comment by nw0** [comment]
Accepted all suggestions.

> Admittedly this is undocumented, we could perhaps add a docs style note somewhere in Contributing.md....

Contributing.md is pretty high-level about documentation -- it's quite clear about the the important points imo. I haven't added a note, but would be happy to write something, or if you want to push an additional thing to this PR.

**Review by davidhewitt — APPROVED** [review]
Perfect, thanks again!

> Contributing.md is pretty high-level about documentation -- it's quite clear about the the important points imo. I haven't added a note, but would be happy to write something, or if you want to push an additional thing to this PR.

Agreed, though also I guess that not many contributors will be writing massive docs PRs once we have everything present. And as we increase docs coverage it's likely that new contributions will match the existing style. Perhaps let's pass on the note for now.

**Pull request merged**